File: src/main/java/OnErrorRetryCache.java
import rx.Observable;

import java.util.concurrent.Semaphore;

public class OnErrorRetryCache<T> {

    public static <T> Observable<T> from(Observable<T> source) {
         return new OnErrorRetryCache<>(source).deferred;
    }

    private final Observable<T> deferred;
    private final Semaphore singlePermit = new Semaphore(1);
    
    private Observable<T> cache = null;
    private Observable<T> inProgress = null;

    private OnErrorRetryCache(Observable<T> source) {
        deferred = Observable.defer(() -> createWhenObserverSubscribes(source));
    }

    private Observable<T> createWhenObserverSubscribes(Observable<T> source) 
    {
        singlePermit.acquireUninterruptibly();
        
        Observable<T> cached = cache;
        if (cached != null) {
            singlePermit.release();
            return cached;
        }

        inProgress = source
                .doOnCompleted(this::onSuccess)
                .doOnTerminate(this::onTermination)
                .replay()
                .autoConnect();

        return inProgress;
    }
    
    private void onSuccess() {
        cache = inProgress;
    }
    
    private void onTermination() {
        inProgress = null;
        singlePermit.release();
    }
}

File: src/main/kotlin/rx/utils/OnErrorRetryCache.kt
package rx.utils

import io.reactivex.Single

import java.util.concurrent.Semaphore

fun <T> Single<T>.cacheButRetryOnError(): Single<T> = OnErrorRetryCache(this).result

private class OnErrorRetryCache<T> constructor(source: Single<T>) {

    private val singlePermit = Semaphore(1)
    private var cache: Single<T>? = null

    val result: Single<T> = Single.defer {
        createOnSubscription(source)
    }

    private fun createOnSubscription(source: Single<T>): Single<T>? {
        singlePermit.acquireUninterruptibly()

        cache?.let {
            singlePermit.release()
            return it
        }

        lateinit var call: Single<T>
        call = source
                .doOnSuccess { cache = call }
                .doAfterTerminate(singlePermit::release)
                .cache()
        return call
    }
}
